### PR TITLE
Editor fixes for custom vertex formats

### PR DESCRIPTION
**Pull request by matgis**
* Changed how we store project data in `graphics_ddf.proto`.
  * The `int_values` and `uint_values` were merged into a single `long_values` field that covers the entire signed and unsigned 32-bit integer range.
  * The `float_values` field was changed to `double_values`. The editor does all math in doubles, and this enables us to add support for double attributes in the future.
  * Normalized values are stored as normalized doubles in `double_values` regardless of binary type. This allows users to change the binary type in the material without breaking overrides in sprites.
  * Regardless of binary type, all value types are converted to `doubles` for editing. Doubles can represent both floats as well as the entire signed and unsigned 32-bit integer range accurately.
* We no longer support pre-provided byte values for VertexAttributes in Bob, because the editor doesn't, and these would be stripped when saving anyway.
* Renamed a bunch of stuff for concistency.
* Fixed the low-level buffer population functions behavior with signed values and moved them to `buffers.clj`.
* Use the info we've already collected in `:attribute-infos` in functions that query shader info from the `ShaderLifecycle`.
* Moved some stuff around between the `vertex2.clj`, `graphics.clj`, and `buffers.clj` modules.
* Invalid vertex attribute definitions in the material now generate build errors.
* Invalid vertex attribute overrides in the sprite now generate build errors and appear red in the Property Editor.
* We now correctly handle different element counts for the `position`, `texcoord`, and `page_index` attributes if they are declared in the material.
* Reordered the vertex attribute columns when editing a material.
* Coerce the vertex attribute values in the material to the new data type and element count when editing their definition.
* Use the `:bytes` we alredy have in the `attribute-info` when building.
* Correctly handle attributes with a single element in the Property Editor.
* Ignore alpha for tree-element color attributes in the Property Editor.
* Hide any material-declared `position`, `texcoord`, and `page_index` attributes of sprites in the Property Editor.
* Disable rendering and show an error in the scene view if we are unable to encode attribute values into the vertex buffer.
* Added tests for number type conversions.
* Added tests for low-level buffer access.
* Added save tests for vertex attributes in materials and sprites.
* Added tests for the `coll.clj` utility module.
* Added a few tests for the `dev.clj` utility module.
* Cleaned up some unrelated things when I saw it.

### CODE REVIEW

File: editor/src/clj/dev.clj [hunk 1]
```diff
@@ -44,7 +44,7 @@
   (ffirst (g/targets-of (workspace) :resource-map)))
 
 (defn app-view []
-  (ffirst (g/targets-of (project) :_node-id)))
+  (ffirst (g/targets-of (project) :selected-node-ids-by-resource-node)))
```
Comment (matgis) [inline editor/src/clj/dev.clj:47]: Unrelated bug fix: The `dev/app-view` function was returning `nil`.

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/GraphicsUtil.java [hunk 2]
```diff
@@ -14,80 +14,62 @@
 
 package com.dynamo.bob.pipeline;
 
-import java.util.List;
-import java.nio.ByteBuffer;
-import java.nio.ByteOrder;
-
```
Comment (matgis) [inline com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/GraphicsUtil.java:20]: Changes in Bob:
* Changed how we store project data in `graphics_ddf.proto`.
  * The `int_values` and `uint_values` were merged into a single `long_values` field that covers the entire signed and unsigned 32-bit integer range.
  * The `float_values` field was changed to `double_values`. The editor does all math in doubles, and this enables us to add support for double attributes in the future.
  * Normalized values are stored as normalized doubles in `double_values` regardless of binary type. This allows users to change the binary type in the material without breaking overrides in sprites.
* We no longer support pre-provided byte values for `VertexAttributes` in Bob, because the editor doesn't, and these would be stripped when saving anyway.

File: editor/src/clj/dev.clj [hunk 3]
```diff
@@ -164,7 +164,9 @@
 (def curve-view (partial view-of-type curve-view/CurveView))
 
 (defn console-view []
-  (-> (view-of-type console/ConsoleNode) (g/targets-of :lines) ffirst))
+  (some-> (view-of-type console/ConsoleNode)
+          (g/targets-of :lines)
+          (ffirst)))
```
Comment (matgis) [inline editor/src/clj/dev.clj:169]: Unrelated bug fix: Would throw instead of returning `nil` if there was no console view.

File: editor/src/clj/editor/buffers.clj [hunk 4]
```diff
@@ -13,8 +13,9 @@
 ;; specific language governing permissions and limitations under the License.
 
 (ns editor.buffers
-  (:import [java.nio Buffer ByteBuffer ByteOrder IntBuffer]
-           [com.google.protobuf ByteString]))
+  (:require [util.num :as num])
+  (:import [com.google.protobuf ByteString]
+           [java.nio ByteBuffer ByteOrder DoubleBuffer FloatBuffer IntBuffer LongBuffer ShortBuffer]))
```
Comment (matgis) [inline editor/src/clj/editor/buffers.clj:18]: * Fixed the low-level buffer population functions behavior with signed values and moved them to `buffers.clj`.

File: editor/src/clj/editor/cljfx_form_view.clj [hunk 5]
```diff
@@ -393,7 +393,7 @@
                     {:text (value->label x)})
      :cell-factory (fn [x]
                      {:text (value->label x)})
-     :items (sort (mapv first options))}))
+     :items (mapv first options)}))
```
Comment (matgis) [inline editor/src/clj/editor/cljfx_form_view.clj:396]: Values in dropdowns were always sorted when editing a material. I kind of want the None values to come first. Also, if you want your items sorted, you should sort them yourself and not rely on the view to do it.

File: editor/src/clj/editor/gl/shader.clj [hunk 6]
```diff
@@ -295,8 +294,7 @@ This must be submitted to the driver for compilation before you can use it. See
   `(def ~name ~(create-shader body)))
 
 (defprotocol ShaderVariables
-  (get-attrib-location [this gl name])
-  (get-attribute-infos [this gl])
+  (attribute-infos [this gl])
```
Comment (matgis) [inline editor/src/clj/editor/gl/shader.clj:297]: We don't like prefixing functions with `get-` in the editor code. 🙂

File: editor/src/clj/editor/gl/vertex.clj [hunk 7]
```diff
@@ -449,27 +448,29 @@ the `do-gl` macro from `editor.gl`."
...
-        norm              (if (not (nil? (first more))) (first more) false)]
-    (when (not= -1 loc)
-      (gl/gl-vertex-attrib-pointer gl ^int loc ^int sz ^int (gl-types tp) ^boolean norm ^int stride ^long offset))))
+  (let [attribute-infos (shader/attribute-infos shader gl)]
+    (mapv (fn [attrib]
+            (let [attribute-name (name (first attrib))]
+              (if-some [attribute-info (get attribute-infos attribute-name)]
+                (:index attribute-info)
+                -1)))
+          attribs)))
```
Comment (matgis) [inline editor/src/clj/editor/gl/vertex.clj:457]: * Use the info we've already collected in `:attribute-infos` in functions that query shader info from the `ShaderLifecycle`.

File: editor/src/clj/editor/gl/vertex2.clj [hunk 8]
```diff
@@ -14,29 +14,19 @@
...
-  (case type
-    :ubyte  Buffers/SIZEOF_BYTE
-    :byte   Buffers/SIZEOF_BYTE
-    :ushort Buffers/SIZEOF_SHORT
-    :short  Buffers/SIZEOF_SHORT
-    :uint   Buffers/SIZEOF_INT
-    :int    Buffers/SIZEOF_INT
-    :float  Buffers/SIZEOF_FLOAT
-    :double Buffers/SIZEOF_DOUBLE))
+(set! *unchecked-math* :warn-on-boxed)
```
Comment (matgis) [inline editor/src/clj/editor/gl/vertex2.clj:29]: * Moved some stuff around between the `vertex2.clj`, `graphics.clj`, and `buffers.clj` modules.

All this stuff is in `buffers.clj` now.

File: editor/src/clj/editor/gl/vertex2.clj [hunk 9]
```diff
@@ -393,14 +207,14 @@
...
-                :ushort `(.putShort  (.shortValue (Long. (bit-and ~arg 0xffff))))
-                :uint   `(.putInt    (.intValue   (Long. (bit-and ~arg 0xffffffff))))))))
+                 :byte   `(.put       ~arg)
+                 :short  `(.putShort  ~arg)
+                 :int    `(.putInt    ~arg)
+                 :float  `(.putFloat  ~arg)
+                 :double `(.putDouble ~arg)
+                 :ubyte  `(.put       (.byteValue  (Long/valueOf (bit-and ~arg 0xff))))
+                 :ushort `(.putShort  (.shortValue (Long/valueOf (bit-and ~arg 0xffff))))
+                 :uint   `(.putInt    (.intValue   (Long/valueOf (bit-and ~arg 0xffffffff))))))))
```
Comment (matgis) [inline editor/src/clj/editor/gl/vertex2.clj:217]: Whitespace + use `valueOf` instead of deprecated `Long` constructor.

File: editor/src/clj/editor/gl/vertex2.clj [hunk 10]
```diff
@@ -451,11 +265,17 @@
...
 (defn- vertex-locate-attribs
   [^GL2 gl shader attribs]
-  (mapv #(shader/get-attrib-location shader gl (:name %)) attribs))
+  (let [attribute-infos (shader/attribute-infos shader gl)]
+    (mapv (fn [attrib]
+            (let [attribute-name (:name attrib)]
+              (if-some [attribute-info (get attribute-infos attribute-name)]
+                (:index attribute-info)
+                -1)))
+          attribs)))
```
Comment (matgis) [inline editor/src/clj/editor/gl/vertex2.clj:274]: * Use the info we've already collected in `:attribute-infos` in functions that query shader info from the `ShaderLifecycle`.

File: editor/src/clj/editor/gl/vertex2.clj [hunk 11]
```diff
@@ -481,10 +301,6 @@
           :when (not= l -1)]
     (gl/gl-disable-vertex-attrib-array gl l)))
 
-(defn- find-attribute-index [attribute-name attributes]
-  (util/first-index-where (fn [attribute] (= attribute-name (:name attribute)))
-                          attributes))
-
```
Comment (matgis) [inline editor/src/clj/editor/gl/vertex2.clj:487]: Unused now.

File: editor/src/clj/editor/gl/vertex2.clj [hunk 12]
```diff
@@ -509,19 +325,19 @@
 
 (defrecord VertexBufferShaderLink [request-id ^VertexBuffer vertex-buffer shader]
   GlBind
-  (bind [_this gl render-args]
+  (bind [_this gl _render-args]
```
Comment (matgis) [inline editor/src/clj/editor/gl/vertex2.clj:328]: Prefixing with and underscore `_` gets rid of the `unused parameter` warnings in IntelliJ.

File: editor/src/clj/editor/sprite.clj [hunk 13]
```diff
@@ -453,91 +496,111 @@
...
-(defn- get-attribute-edit-type [attribute prop-type]
-  (let [attribute-semantic-type (:semantic-type attribute)
+(defn- attribute-edit-type [attribute property-type]
+  (let [attribute-element-count (:element-count attribute)
+        attribute-semantic-type (:semantic-type attribute)
         attribute-update-fn (fn [_evaluation-context self _old-value new-value]
-                              (g/update-property self :vertex-attribute-overrides attribute-update-property attribute new-value))
+                              (let [values (if (= g/Num property-type)
+                                             (vector-of :double new-value)
+                                             new-value)]
```
Comment (matgis) [inline editor/src/clj/editor/sprite.clj:528]: The `g/Num` property type expects a number, so we must unpack single-element attributes from their vector for editing using the property editor widget.

File: editor/src/clj/editor/sprite.clj [hunk 14]
```diff
@@ -453,91 +496,111 @@
...
+(def ^:private attribute-key->property-key (memoize attribute-key->property-key-raw))
+
+(def ^:private attribute-info->property-key (comp attribute-key->property-key :name-key))
 
-(defn- attribute->outline-key [attribute-info]
-  (keyword (str "attribute_" (name (:name-key attribute-info)))))
+(defn- editable-attribute-info? [attribute-info]
+  (case (:semantic-type attribute-info)
+    (:semantic-type-position :semantic-type-texcoord :semantic-type-page-index) false
+    true))
```
Comment (matgis) [inline editor/src/clj/editor/sprite.clj:549]: * Hide any material-declared `position`, `texcoord`, and `page_index` attributes of sprites in the Property Editor.

File: editor/src/clj/editor/sprite.clj [hunk 15]
```diff
@@ -453,91 +496,111 @@
...
+                              :type property-type
+                              :edit-type edit-type
+                              :label label
+                              :value value
+                              :error error}]
+                    ;; Insert the original material values as original-value if there is a vertex override.
+                    (if (some? override-values)
+                      [property-key (assoc prop :original-value material-values)]
+                      [property-key prop]))))
+              material-attribute-infos)]
```
Comment (matgis) [inline editor/src/clj/editor/sprite.clj:580]: Mostly whitespace + renames, but we also decorate invalid attributes with errors to make them red and get an informative tooltip in the property editor.

File: editor/src/clj/editor/sprite.clj [hunk 16]
```diff
@@ -453,91 +496,111 @@
...
+                                                 [bytes error-message :as bytes+error-message] (graphics/attribute->bytes+error-message attribute-info resized-values)]
+                                             (if (nil? error-message)
+                                               bytes+error-message
+                                               (let [property-key (attribute-key->property-key name-key)
+                                                     error (g/->error _node-id property-key :fatal override-values error-message)]
+                                                 [bytes error]))))]
+                       [name-key (or error bytes)])))
+              material-attribute-infos)]
+    (g/precluding-errors (vals vertex-attribute-bytes)
+      vertex-attribute-bytes)))
```
Comment (matgis) [inline editor/src/clj/editor/sprite.clj:603]: Again, we can fail to encode the values and need to propagate the errors to elsewhere somehow.

File: editor/src/clj/editor/sprite.clj [hunk 17]
```diff
@@ -621,10 +684,6 @@
             (default {})
             (dynamic visible (g/constantly false)))
 
-  (property attributes g/Any
-            (default [])
-            (dynamic visible (g/constantly false)))
-
```
Comment (matgis) [inline editor/src/clj/editor/sprite.clj:627]: This was unused. We store only the override values by attribute-key in the `vertex-attribute-overrides` property.

File: editor/src/clj/editor/validation.clj [hunk 18]
```diff
@@ -87,7 +87,7 @@
                "'%s' must be between %d and %d"
                "'%s' must be between %f and %f")]
     (when (not (<= min v max))
-      (format tmpl name min max))))
+      (util/format* tmpl name min max))))
```
Comment (matgis) [inline editor/src/clj/editor/validation.clj:90]: `util/format*` uses the `ROOT` locale instead of the system locale to ensure float values have decimal points and not commas.

File: editor/test/dev_test.clj [hunk 19]
```diff
@@ -0,0 +1,29 @@
...
+            [integration.test-util :as test-util]))
+
+(set! *warn-on-reflection* true)
+
+(deftest node-type-accessors-test
+  (test-util/with-loaded-project "test/resources/empty_project"
+    (is (= :editor.workspace/Workspace (g/node-type-kw (dev/workspace))))
+    (is (= :editor.defold-project/Project (g/node-type-kw (dev/project))))
+    (is (= (project/workspace (dev/project)) (dev/workspace)))
+    (is (= app-view (dev/app-view)))))
```
Comment (matgis) [inline editor/test/dev_test.clj:29]: I ran into an issue with the `dev/app-view` function when testing, so I added a test to make sure it stays working from now on. 😅

File: editor/test/integration/animation_set_test.clj [hunk 20]
```diff
@@ -56,7 +56,7 @@
           (testing "Channels are not animated by multiple tracks"
             (doseq [[channel data] data-by-channel]
               (is (not= ::conflicting-data data)
-                  (str "Found multiple tracks targetting " channel " for bone " bone-id))))
+                  (str "Found multiple tracks targeting " channel " for bone " bone-id))))
```
Comment (matgis) [inline editor/test/integration/animation_set_test.clj:59]: Unrelated typo.

File: editor/src/clj/editor/graphics.clj [hunk 21]
```diff
@@ -13,90 +13,198 @@
...
+(defn validate-doubles [double-values attribute node-id prop-kw]
+  (validation/prop-error :fatal node-id prop-kw doubles-outside-attribute-range-error-message double-values attribute))
 
-(defn make-attribute-byte-buffer
-  ^ByteBuffer [attribute-data-type attribute-values]
+(defn resize-doubles [double-values semantic-type element-count]
+  (let [fill-value (case semantic-type
+                     :semantic-type-color 1.0 ; Default to opaque white for color attributes.
+                     0.0)]
+    (coll/resize double-values element-count fill-value)))
```
[the remark this answers is not in the document]
Comment (Jhonnyg) [reply editor/src/clj/editor/graphics.clj:125]: Yes, I agree

**Review by Jhonnyg — APPROVED** [review]
Great stuff!
